## Exercise L3 - 1: Diagnose Dataset Level and Select Last Encounter

### Instructions
- Given the dataset, convert the dataset to a longitudinal level but select only the last encounter for each patient.
- Assume that that the order of encounter IDs is indicative of the time for encounter. In other words a lower number encounter will come before a higher numbered encounter.

In [2]:
import pandas as pd
import numpy as np

In [3]:
ehr_level_dataset_path = "./data/ehr_level_exercise_dataset.csv"

### Level of Dataset
What level is the dataset at? Is at the line or encounter level? 

### Solution

In [4]:
ehr_level_df = pd.read_csv(ehr_level_dataset_path)
ehr_level_df.head()

ENCOUNTER_ID                     PATIENT_ID  \
0     udacity_health_encounter_id_1  udacity_health_patient_id_186   
1    udacity_health_encounter_id_10  udacity_health_patient_id_188   
2   udacity_health_encounter_id_100  udacity_health_patient_id_585   
3  udacity_health_encounter_id_1000  udacity_health_patient_id_525   
4  udacity_health_encounter_id_1001  udacity_health_patient_id_950   

  PRINCIPAL_DIAGNOSIS_CODE                                     PROCEDURE_CODE  \
0            dx_code_15406  ['procedure_code_58552', 'procedure_code_39776...   
1            dx_code_74047                                                 []   
2            dx_code_71465                                                 []   
3            dx_code_61569                                                 []   
4            dx_code_90172                           ['procedure_code_30555']   

                                     MEDICATION_CODE  \
0  ['medication_code_2350', 'medication_code_8630...   
1  ['medication_code_7789', 'medication_code_3560...   
2  ['medication_code_7982', 'medication_code_2452...   
3                           ['medication_code_4036']   
4   ['medication_code_6755', 'medication_code_5045']   

                                            LAB_CODE  \
0  ['lab_code_8835', 'lab_code_9859', 'lab_code_9...   
1                                                 []   
2  ['lab_code_4198', 'lab_code_6603', 'lab_code_3...   
3                                                 []   
4                                  ['lab_code_9112']   

                                               LABEL  
0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  
1                                       [1, 1, 1, 1]  
2                              [0, 0, 0, 0, 0, 0, 0]  
3                                                [0]  
4                                       [0, 0, 0, 0]

In [5]:
ehr_level_df.to_csv("data.csv")

In [6]:
ls

data/
data.csv
Lesson_EHR_Data_Transformations_Feature_Engineering_exercises_solutions.ipynb
vocab/


**Tests**
- Line: Total number of rows > Number of Unique Encounters
- Encounter level: Total Number of Rows = Number of Unique Encounters


In [7]:
# Line Test
try:
    assert len(ehr_level_df) > ehr_level_df['ENCOUNTER_ID'].nunique() 
    print("Dataset could be at the line level")
except:
    print("Dataset is not at the line level")

Dataset is not at the line level


In [8]:
# Encounter Test
try:
    assert len(ehr_level_df) == ehr_level_df['ENCOUNTER_ID'].nunique()
    print("Dataset could be at the encounter level")
except:
    print("Dataset is not at the encounter level")

Dataset could be at the encounter level


**Answer:** Dataset is at the encounter level and you can probably guess by seeing the arrays for the code sets but we did a few simple tests to confirm.

### Select Last Encounter for each Patient

So in many cases you may only want a snapshot of a patient's history for your modeling objective. In some cases it might be important to see the changes over time but in other cases you only want the most recent case or depending on the model the first case could also be used. Really important to know how the context for how the model will be deployed in production and the time state you will be getting data.

In [9]:
ehr_level_df.head()

ENCOUNTER_ID                     PATIENT_ID  \
0     udacity_health_encounter_id_1  udacity_health_patient_id_186   
1    udacity_health_encounter_id_10  udacity_health_patient_id_188   
2   udacity_health_encounter_id_100  udacity_health_patient_id_585   
3  udacity_health_encounter_id_1000  udacity_health_patient_id_525   
4  udacity_health_encounter_id_1001  udacity_health_patient_id_950   

  PRINCIPAL_DIAGNOSIS_CODE                                     PROCEDURE_CODE  \
0            dx_code_15406  ['procedure_code_58552', 'procedure_code_39776...   
1            dx_code_74047                                                 []   
2            dx_code_71465                                                 []   
3            dx_code_61569                                                 []   
4            dx_code_90172                           ['procedure_code_30555']   

                                     MEDICATION_CODE  \
0  ['medication_code_2350', 'medication_code_8630...   
1  ['medication_code_7789', 'medication_code_3560...   
2  ['medication_code_7982', 'medication_code_2452...   
3                           ['medication_code_4036']   
4   ['medication_code_6755', 'medication_code_5045']   

                                            LAB_CODE  \
0  ['lab_code_8835', 'lab_code_9859', 'lab_code_9...   
1                                                 []   
2  ['lab_code_4198', 'lab_code_6603', 'lab_code_3...   
3                                                 []   
4                                  ['lab_code_9112']   

                                               LABEL  
0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  
1                                       [1, 1, 1, 1]  
2                              [0, 0, 0, 0, 0, 0, 0]  
3                                                [0]  
4                                       [0, 0, 0, 0]

In [10]:
ehr_level_df.tail(1)

ENCOUNTER_ID                     PATIENT_ID  \
7221  udacity_health_encounter_id_999  udacity_health_patient_id_547   

     PRINCIPAL_DIAGNOSIS_CODE  \
7221            dx_code_83632   

                                         PROCEDURE_CODE  \
7221  ['procedure_code_70874', 'procedure_code_38557...   

                                        MEDICATION_CODE  \
7221  ['medication_code_3401', 'medication_code_6448...   

                                               LAB_CODE  \
7221  ['lab_code_1957', 'lab_code_6875', 'lab_code_8...   

                                              LABEL  
7221  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

In [11]:
ehr_level_df.groupby('PATIENT_ID').count()

ENCOUNTER_ID  PRINCIPAL_DIAGNOSIS_CODE  \
PATIENT_ID                                                               
udacity_health_patient_id_1                6                         6   
udacity_health_patient_id_10              30                        30   
udacity_health_patient_id_100             14                        14   
udacity_health_patient_id_1000            13                        13   
udacity_health_patient_id_101              2                         2   
...                                      ...                       ...   
udacity_health_patient_id_995              4                         4   
udacity_health_patient_id_996              2                         2   
udacity_health_patient_id_997              1                         1   
udacity_health_patient_id_998              9                         9   
udacity_health_patient_id_999              2                         2   

                                PROCEDURE_CODE  MEDICATION_CODE  LAB_CODE  \
PATIENT_ID                                                                  
udacity_health_patient_id_1                  6                6         6   
udacity_health_patient_id_10                30               30        30   
udacity_health_patient_id_100               14               14        14   
udacity_health_patient_id_1000              13               13        13   
udacity_health_patient_id_101                2                2         2   
...                                        ...              ...       ...   
udacity_health_patient_id_995                4                4         4   
udacity_health_patient_id_996                2                2         2   
udacity_health_patient_id_997                1                1         1   
udacity_health_patient_id_998                9                9         9   
udacity_health_patient_id_999                2                2         2   

                                LABEL  
PATIENT_ID                             
udacity_health_patient_id_1         6  
udacity_health_patient_id_10       30  
udacity_health_patient_id_100      14  
udacity_health_patient_id_1000     13  
udacity_health_patient_id_101       2  
...                               ...  
udacity_health_patient_id_995       4  
udacity_health_patient_id_996       2  
udacity_health_patient_id_997       1  
udacity_health_patient_id_998       9  
udacity_health_patient_id_999       2  

[878 rows x 6 columns]

In [12]:
ehr_level_df.groupby('PATIENT_ID')["ENCOUNTER_ID"].count()

PATIENT_ID
udacity_health_patient_id_1        6
udacity_health_patient_id_10      30
udacity_health_patient_id_100     14
udacity_health_patient_id_1000    13
udacity_health_patient_id_101      2
                                  ..
udacity_health_patient_id_995      4
udacity_health_patient_id_996      2
udacity_health_patient_id_997      1
udacity_health_patient_id_998      9
udacity_health_patient_id_999      2
Name: ENCOUNTER_ID, Length: 878, dtype: int64

In [13]:
ehr_level_df.groupby('PATIENT_ID')["ENCOUNTER_ID"].tail(1)

20      udacity_health_encounter_id_1018
27      udacity_health_encounter_id_1026
46      udacity_health_encounter_id_1044
66      udacity_health_encounter_id_1062
77      udacity_health_encounter_id_1074
                      ...               
7217     udacity_health_encounter_id_995
7218     udacity_health_encounter_id_996
7219     udacity_health_encounter_id_997
7220     udacity_health_encounter_id_998
7221     udacity_health_encounter_id_999
Name: ENCOUNTER_ID, Length: 878, dtype: object

In [14]:
# select last encounter for each patient
#convert encounter id column to a numerical value
def convert_encounter_id_to_number(df, encounter_id):
    df["ENCOUNTER_ID_NUMBER"] = df[encounter_id].str.replace('udacity_health_encounter_id_', '').astype(int)
    return df

def select_last_encounter(df, patient_id, encounter_id):
    df = df.sort_values(encounter_id)
    last_encounter_values = df.groupby(patient_id)[encounter_id].tail(1).values
    return df[df[encounter_id].isin(last_encounter_values)] 

In [15]:
ehr_encounter_number_df = convert_encounter_id_to_number(ehr_level_df, "ENCOUNTER_ID")
last_encounter_df = select_last_encounter(ehr_encounter_number_df, "PATIENT_ID", "ENCOUNTER_ID_NUMBER" )

In [16]:
#take subset of output
test_last_encounter_df = last_encounter_df[['ENCOUNTER_ID', 'ENCOUNTER_ID_NUMBER', 'PATIENT_ID']]

### Test cases 
- PATIENT_IDS - udacity_health_patient_id_309, udacity_health_patient_id_418, udacity_health_patient_id_908

In [17]:
ehr_level_df[ehr_level_df['PATIENT_ID']=='udacity_health_patient_id_309']

ENCOUNTER_ID                     PATIENT_ID  \
94    udacity_health_encounter_id_1091  udacity_health_patient_id_309   
450   udacity_health_encounter_id_1436  udacity_health_patient_id_309   
605   udacity_health_encounter_id_1582  udacity_health_patient_id_309   
665   udacity_health_encounter_id_1643  udacity_health_patient_id_309   
684   udacity_health_encounter_id_1662  udacity_health_patient_id_309   
867   udacity_health_encounter_id_1837  udacity_health_patient_id_309   
1024  udacity_health_encounter_id_1985  udacity_health_patient_id_309   
1042  udacity_health_encounter_id_2000  udacity_health_patient_id_309   
1196  udacity_health_encounter_id_2142  udacity_health_patient_id_309   
1289  udacity_health_encounter_id_2229  udacity_health_patient_id_309   
1340  udacity_health_encounter_id_2281  udacity_health_patient_id_309   
1596  udacity_health_encounter_id_2531  udacity_health_patient_id_309   
1702  udacity_health_encounter_id_2631  udacity_health_patient_id_309   
2069  udacity_health_encounter_id_2989  udacity_health_patient_id_309   
2291  udacity_health_encounter_id_3209  udacity_health_patient_id_309   
2507  udacity_health_encounter_id_3416  udacity_health_patient_id_309   
2696   udacity_health_encounter_id_360  udacity_health_patient_id_309   
2890  udacity_health_encounter_id_3792  udacity_health_patient_id_309   
2960  udacity_health_encounter_id_3859  udacity_health_patient_id_309   
3064  udacity_health_encounter_id_3961  udacity_health_patient_id_309   
3167  udacity_health_encounter_id_4058  udacity_health_patient_id_309   
3280  udacity_health_encounter_id_4168  udacity_health_patient_id_309   
3353  udacity_health_encounter_id_4239  udacity_health_patient_id_309   
3561  udacity_health_encounter_id_4439  udacity_health_patient_id_309   
3585  udacity_health_encounter_id_4462  udacity_health_patient_id_309   
3718  udacity_health_encounter_id_4589  udacity_health_patient_id_309   
3752  udacity_health_encounter_id_4619  udacity_health_patient_id_309   
3916   udacity_health_encounter_id_478  udacity_health_patient_id_309   
4552  udacity_health_encounter_id_5398  udacity_health_patient_id_309   
4687   udacity_health_encounter_id_553  udacity_health_patient_id_309   
4771  udacity_health_encounter_id_5611  udacity_health_patient_id_309   
4785  udacity_health_encounter_id_5624  udacity_health_patient_id_309   
4796  udacity_health_encounter_id_5635  udacity_health_patient_id_309   
4862  udacity_health_encounter_id_5701  udacity_health_patient_id_309   
4881  udacity_health_encounter_id_5721  udacity_health_patient_id_309   
5132   udacity_health_encounter_id_597  udacity_health_patient_id_309   
5306  udacity_health_encounter_id_6139  udacity_health_patient_id_309   
5385   udacity_health_encounter_id_622  udacity_health_patient_id_309   
5472  udacity_health_encounter_id_6305  udacity_health_patient_id_309   
5564  udacity_health_encounter_id_6394  udacity_health_patient_id_309   
5669  udacity_health_encounter_id_6497  udacity_health_patient_id_309   
5684  udacity_health_encounter_id_6512  udacity_health_patient_id_309   
5829  udacity_health_encounter_id_6662  udacity_health_patient_id_309   
5846   udacity_health_encounter_id_668  udacity_health_patient_id_309   
5946  udacity_health_encounter_id_6777  udacity_health_patient_id_309   
6508  udacity_health_encounter_id_7327  udacity_health_patient_id_309   
6892  udacity_health_encounter_id_7696  udacity_health_patient_id_309   
6969  udacity_health_encounter_id_7772  udacity_health_patient_id_309   
7217   udacity_health_encounter_id_995  udacity_health_patient_id_309   

     PRINCIPAL_DIAGNOSIS_CODE  \
94              dx_code_62502   
450             dx_code_37825   
605             dx_code_79663   
665             dx_code_60773   
684             dx_code_29272   
867             dx_code_56368   
1024            dx_code_38581   
1042            dx_code_91230   
1196            dx_code_58527   
1289            dx_code_46826   
1340        

For patient id 309, the selected encounter should be 7772.

In [18]:
test_last_encounter_df[test_last_encounter_df['PATIENT_ID']=='udacity_health_patient_id_309']

ENCOUNTER_ID  ENCOUNTER_ID_NUMBER  \
6969  udacity_health_encounter_id_7772                 7772   

                         PATIENT_ID  
6969  udacity_health_patient_id_309

In [19]:
ehr_level_df[ehr_level_df['PATIENT_ID']=='udacity_health_patient_id_418']

ENCOUNTER_ID                     PATIENT_ID  \
263   udacity_health_encounter_id_1258  udacity_health_patient_id_418   
1230  udacity_health_encounter_id_2173  udacity_health_patient_id_418   
2448  udacity_health_encounter_id_3362  udacity_health_patient_id_418   

     PRINCIPAL_DIAGNOSIS_CODE  \
263             dx_code_87320   
1230            dx_code_72104   
2448            dx_code_77979   

                                         PROCEDURE_CODE  \
263                                                  []   
1230  ['procedure_code_37393', 'procedure_code_48757...   
2448                           ['procedure_code_64397']   

                                        MEDICATION_CODE  \
263                            ['medication_code_7475']   
1230  ['medication_code_3809', 'medication_code_9141...   
2448   ['medication_code_2807', 'medication_code_6979']   

                                               LAB_CODE  \
263                                                  []   
1230  ['lab_code_2839', 'lab_code_2665', 'lab_code_3...   
2448                                                 []   

                                                  LABEL  ENCOUNTER_ID_NUMBER  
263                                                 [0]                 1258  
1230  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...                 2173  
2448                                          [0, 0, 0]                 3362

For patient id 418, the selected encounter should be 3362.

In [20]:
test_last_encounter_df[test_last_encounter_df['PATIENT_ID']=='udacity_health_patient_id_418']

ENCOUNTER_ID  ENCOUNTER_ID_NUMBER  \
2448  udacity_health_encounter_id_3362                 3362   

                         PATIENT_ID  
2448  udacity_health_patient_id_418

In [21]:
ehr_level_df[ehr_level_df['PATIENT_ID']=='udacity_health_patient_id_908']

ENCOUNTER_ID                     PATIENT_ID  \
1678  udacity_health_encounter_id_2608  udacity_health_patient_id_908   
2796  udacity_health_encounter_id_3698  udacity_health_patient_id_908   
5301  udacity_health_encounter_id_6132  udacity_health_patient_id_908   
7131   udacity_health_encounter_id_910  udacity_health_patient_id_908   

     PRINCIPAL_DIAGNOSIS_CODE  \
1678            dx_code_52877   
2796            dx_code_52688   
5301            dx_code_36066   
7131            dx_code_98539   

                                         PROCEDURE_CODE  \
1678                                                 []   
2796                           ['procedure_code_60912']   
5301  ['procedure_code_5724', 'procedure_code_55191'...   
7131  ['procedure_code_8188', 'procedure_code_59872'...   

                                        MEDICATION_CODE  \
1678                           ['medication_code_6189']   
2796                                                 []   
5301  ['medication_code_9532', 'medication_code_3398...   
7131  ['medication_code_5112', 'medication_code_3537...   

                                               LAB_CODE  \
1678                                                 []   
2796                 ['lab_code_4691', 'lab_code_9380']   
5301  ['lab_code_9732', 'lab_code_2808', 'lab_code_3...   
7131                                                 []   

                                           LABEL  ENCOUNTER_ID_NUMBER  
1678                                         [0]                 2608  
2796                                   [0, 0, 0]                 3698  
5301  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]                 6132  
7131                       [0, 0, 0, 0, 0, 0, 0]                  910

For patient id 908, the selected encounter should be 6132.

In [22]:
test_last_encounter_df[test_last_encounter_df['PATIENT_ID']=='udacity_health_patient_id_908']

ENCOUNTER_ID  ENCOUNTER_ID_NUMBER  \
5301  udacity_health_encounter_id_6132                 6132   

                         PATIENT_ID  
5301  udacity_health_patient_id_908

## Exercise  L3 - 2: Dataset Splitting

### Instructions
- Split the provided dataset into a train and test split but be sure not to mix patient encounter records across the two partitions
- Be sure to run the following three tests
    - Patient data in only one partition
    - Total unique number of patients across all partitions = total number unique patients in original full dataset
    - Total number of rows original dataset = sum of rows across splits

In [23]:
splitting_exercise_dataset_path = "./data/SYNTHETIC_EHR_DATASET.csv"

### Solution

This is largely a review of two parts in this lesson and you can use most of the same code for each step. The key is to identify the level of the dataset and then to convert it to the encounter level before you do your splits. Then perform the splitting and run the tests.

#### Convert to Encounter Level

In [24]:
# convert to encounter and then split but make sure not to mix patient encounter records across the two partitions
ehr_pre_split_df = pd.read_csv(splitting_exercise_dataset_path)
print("Rows x Columns in ehr_pre_split_df = ", ehr_pre_split_df.shape)

Rows x Columns in ehr_pre_split_df =  (100000, 7)


In [25]:
ehr_pre_split_df.head(2)

ENCOUNTER_ID                     PATIENT_ID  \
0  udacity_health_encounter_id_1933  udacity_health_patient_id_817   
1  udacity_health_encounter_id_5664  udacity_health_patient_id_594   

  PRINCIPAL_DIAGNOSIS_CODE        PROCEDURE_CODE       MEDICATION_CODE  \
0            dx_code_56522  procedure_code_20005                   NaN   
1            dx_code_39264                   NaN  medication_code_7471   

  LAB_CODE  LABEL  
0      NaN      0  
1      NaN      0

In [26]:
ehr_pre_split_df.columns

Index(['ENCOUNTER_ID', 'PATIENT_ID', 'PRINCIPAL_DIAGNOSIS_CODE',
       'PROCEDURE_CODE', 'MEDICATION_CODE', 'LAB_CODE', 'LABEL'],
      dtype='object')

In [27]:
print([x for x in range(51) if x%2==0])

[0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50]


In [28]:
print([x for x in range(51) if x%2==0 and x%3==0 and x%5==0 and x>10])

[30]


In [29]:
print([x+y for x in range(10) for y in range (10) if (x+y)%10==0])

[0, 10, 10, 10, 10, 10, 10, 10, 10, 10]


In [30]:
grouping_field_list = ['ENCOUNTER_ID', 'PATIENT_ID', 'PRINCIPAL_DIAGNOSIS_CODE']

non_grouped_field_list = [c for c in ehr_pre_split_df.columns if c not in grouping_field_list]

In [31]:
[c for c in ehr_pre_split_df.columns if c not in grouping_field_list]

['PROCEDURE_CODE', 'MEDICATION_CODE', 'LAB_CODE', 'LABEL']

In [32]:
ehr_pre_split_df.groupby(grouping_field_list).count()

PROCEDURE_CODE  \
ENCOUNTER_ID                     PATIENT_ID                    PRINCIPAL_DIAGNOSIS_CODE                   
udacity_health_encounter_id_1    udacity_health_patient_id_186 dx_code_15406                          4   
udacity_health_encounter_id_10   udacity_health_patient_id_188 dx_code_74047                          0   
udacity_health_encounter_id_100  udacity_health_patient_id_585 dx_code_71465                          0   
udacity_health_encounter_id_1000 udacity_health_patient_id_525 dx_code_61569                          0   
udacity_health_encounter_id_1001 udacity_health_patient_id_950 dx_code_90172                          1   
...                                                                                                 ...   
udacity_health_encounter_id_995  udacity_health_patient_id_309 dx_code_4976                           2   
udacity_health_encounter_id_996  udacity_health_patient_id_692 dx_code_3150                           0   
udacity_health_encounter_id_997  udacity_health_patient_id_9   dx_code_82423                          1   
udacity_health_encounter_id_998  udacity_health_patient_id_477 dx_code_14970                          5   
udacity_health_encounter_id_999  udacity_health_patient_id_547 dx_code_83632                          5   

                                                                                         MEDICATION_CODE  \
ENCOUNTER_ID                     PATIENT_ID                    PRINCIPAL_DIAGNOSIS_CODE                    
udacity_health_encounter_id_1    udacity_health_patient_id_186 dx_code_15406                          13   
udacity_health_encounter_id_10   udacity_health_patient_id_188 dx_code_74047                           4   
udacity_health_encounter_id_100  udacity_health_patient_id_585 dx_code_71465                           3   
udacity_health_encounter_id_1000 udacity_health_patient_id_525 dx_code_61569                           1   
udacity_health_encounter_id_1001 udacity_health_patient_id_950 dx_code_90172                           2   
...                                                                                                  ...   
udacity_health_encounter_id_995  udacity_health_patient_id_309 dx_code_4976                            4   
udacity_health_encounter_id_996  udacity_health_patient_id_692 dx_code_3150                            8   
udacity_health_encounter_id_997  udacity_health_patient_id_9   dx_code_82423                           4   
udacity_health_encounter_id_998  udacity_health_patient_id_477 dx_code_14970                          16   
udacity_health_encounter_id_999  udacity_health_patient_id_547 dx_code_83632                           7   

                                                                                         LAB_CODE  \
ENCOUNTER_ID                     PATIENT_ID                    PRINCIPAL_DIAGNOSIS_CODE             
udacity_health_encounter_id_1    udacity_health_patient_id_186 dx_code_15406                    4   
udacity_health_encounter_id_10   udacity_health_patient_id_188 dx_code_74047                    0   
udacity_health_encounter_id_100  udacity_health_patient_id_585 dx_code_71465                    4   
udacity_health_encounter_id_1000 udacity_health_patient_id_525 dx_code_61569                    0   
udacity_health_encounter_id_1001 udacity_health_patient_id_950 dx_code_90172                    1   
...                                                                                           ...   
udacity_health_encounter_id_995  udacity_health_patient_id_309 dx_code_4976                     1   
udacity_health_encounter_id_996  udacity_health_patient_id_692 dx_code_3150                     1   
udacity_health_encounter_id_997  udacity_health_patient_id_9   dx_code_82423                    0   
udacity_health_encounter_id_998  udacity_health_patient_id_477 dx_code_14970                    5   
udacity_health_encounter_id_999  udacity_health_patient_id_547 dx_code_83632    

In [33]:
DF = pd.DataFrame({"A": np.random.randint(1,100,1000),\
             "B": np.random.randint(1,100,1000),\
             "C": np.random.randint(1,100,1000),\
             "D": np.random.randint(1,100,1000),\
              "N": [np.random.choice(["Ram","Shyam","sita","gita","hari"]) for x in range(1000)] })

In [34]:
DF.head()

A   B   C   D      N
0  37  10  43  55   hari
1  62  98  79  81  Shyam
2  56  73  25  49  Shyam
3  98  28   8  33    Ram
4  21  79  40  57   gita

In [35]:
def dosomthing(x):
    return x*x + x+10

In [36]:
col =['A','B','C','D']

In [37]:
DF1 = DF[col].apply(dosomthing)
DF1.head()

A     B     C     D
0  1416   120  1902  3090
1  3916  9712  6330  6652
2  3202  5412   660  2460
3  9712   822    82  1132
4   472  6330  1650  3316

In [38]:
DF2 =DF[col].apply(lambda x: x*x+x+10)
DF2.head()

A     B     C     D
0  1416   120  1902  3090
1  3916  9712  6330  6652
2  3202  5412   660  2460
3  9712   822    82  1132
4   472  6330  1650  3316

In [39]:
DF3 = DF.groupby("N").count()
DF3.head()

A    B    C    D
N                        
Ram    201  201  201  201
Shyam  195  195  195  195
gita   200  200  200  200
hari   203  203  203  203
sita   201  201  201  201

In [40]:
DF3 = DF.groupby("N")[col].agg(lambda x : sum(x))
DF3.head()

A      B      C      D
N                                
Ram     9826  10194  10558  10235
Shyam  10068   9977   9751   9317
gita   10594  10584   9829   9651
hari   10111  10291   9996  10313
sita    9521  10472   9898   9927

In [41]:
DF4 = DF[DF['N']=='Ram']
DF4.head()

A   B   C   D    N
3   98  28   8  33  Ram
13  55  27  36  41  Ram
19  68  61  47   3  Ram
23   3  86  74  75  Ram
27  26  40  62  16  Ram

In [42]:
DF4 = DF[DF['N'].isin(['Ram','Shyam'])]
DF4.head()

A   B   C   D      N
1  62  98  79  81  Shyam
2  56  73  25  49  Shyam
3  98  28   8  33    Ram
5  60  35  56   7  Shyam
8  87  90  20  22  Shyam

In [43]:
DF4.set_index('N')

A   B   C   D
N                    
Shyam  62  98  79  81
Shyam  56  73  25  49
Ram    98  28   8  33
Shyam  60  35  56   7
Shyam  87  90  20  22
...    ..  ..  ..  ..
Ram    55   1   4  73
Shyam  99  12  79  21
Ram    23  60  94  84
Shyam  71  56  80  76
Shyam  80  89  79  29

[396 rows x 4 columns]

In [44]:
DF4.loc[10:15,"A":"B"]

A   B
11  19  82
13  55  27
15  19  61

In [45]:
#DF4.loc[10,"A"]

In [46]:
DF4.groupby(['N']).count()

A    B    C    D
N                        
Ram    201  201  201  201
Shyam  195  195  195  195

In [47]:
ehr_pre_split_df.groupby(grouping_field_list)[non_grouped_field_list].agg(lambda x: list(x ) ).head(1)

PROCEDURE_CODE  \
ENCOUNTER_ID                  PATIENT_ID                    PRINCIPAL_DIAGNOSIS_CODE                                                      
udacity_health_encounter_id_1 udacity_health_patient_id_186 dx_code_15406             [nan, nan, nan, procedure_code_58552, nan, pro...   

                                                                                                                        MEDICATION_CODE  \
ENCOUNTER_ID                  PATIENT_ID                    PRINCIPAL_DIAGNOSIS_CODE                                                      
udacity_health_encounter_id_1 udacity_health_patient_id_186 dx_code_15406             [medication_code_2350, medication_code_8630, n...   

                                                                                                                               LAB_CODE  \
ENCOUNTER_ID                  PATIENT_ID                    PRINCIPAL_DIAGNOSIS_CODE                                                      
udacity_health_encounter_id_1 udacity_health_patient_id_186 dx_code_15406             [nan, nan, lab_code_8835, nan, nan, nan, nan, ...   

                                                                                                                                  LABEL  
ENCOUNTER_ID                  PATIENT_ID                    PRINCIPAL_DIAGNOSIS_CODE                                                     
udacity_health_encounter_id_1 udacity_health_patient_id_186 dx_code_15406             [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...

In [48]:
ehr_pre_split_df.groupby(grouping_field_list)[non_grouped_field_list].agg(lambda x: 
                                                        list([y for y in x if y is not np.nan ] ) ).head(2)

PROCEDURE_CODE  \
ENCOUNTER_ID                   PATIENT_ID                    PRINCIPAL_DIAGNOSIS_CODE                                                      
udacity_health_encounter_id_1  udacity_health_patient_id_186 dx_code_15406             [procedure_code_58552, procedure_code_39776, p...   
udacity_health_encounter_id_10 udacity_health_patient_id_188 dx_code_74047                                                            []   

                                                                                                                         MEDICATION_CODE  \
ENCOUNTER_ID                   PATIENT_ID                    PRINCIPAL_DIAGNOSIS_CODE                                                      
udacity_health_encounter_id_1  udacity_health_patient_id_186 dx_code_15406             [medication_code_2350, medication_code_8630, m...   
udacity_health_encounter_id_10 udacity_health_patient_id_188 dx_code_74047             [medication_code_7789, medication_code_3560, m...   

                                                                                                                                LAB_CODE  \
ENCOUNTER_ID                   PATIENT_ID                    PRINCIPAL_DIAGNOSIS_CODE                                                      
udacity_health_encounter_id_1  udacity_health_patient_id_186 dx_code_15406             [lab_code_8835, lab_code_9859, lab_code_9032, ...   
udacity_health_encounter_id_10 udacity_health_patient_id_188 dx_code_74047                                                            []   

                                                                                                                                   LABEL  
ENCOUNTER_ID                   PATIENT_ID                    PRINCIPAL_DIAGNOSIS_CODE                                                     
udacity_health_encounter_id_1  udacity_health_patient_id_186 dx_code_15406             [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  
udacity_health_encounter_id_10 udacity_health_patient_id_188 dx_code_74047                                                  [1, 1, 1, 1]

In [49]:
ehr_encounter_df = ehr_pre_split_df.groupby(grouping_field_list)[non_grouped_field_list].agg(lambda x: 
                                                        list([y for y in x if y is not np.nan ] ) ).reset_index()

In [50]:
ehr_encounter_df.head()

ENCOUNTER_ID                     PATIENT_ID  \
0     udacity_health_encounter_id_1  udacity_health_patient_id_186   
1    udacity_health_encounter_id_10  udacity_health_patient_id_188   
2   udacity_health_encounter_id_100  udacity_health_patient_id_585   
3  udacity_health_encounter_id_1000  udacity_health_patient_id_525   
4  udacity_health_encounter_id_1001  udacity_health_patient_id_950   

  PRINCIPAL_DIAGNOSIS_CODE                                     PROCEDURE_CODE  \
0            dx_code_15406  [procedure_code_58552, procedure_code_39776, p...   
1            dx_code_74047                                                 []   
2            dx_code_71465                                                 []   
3            dx_code_61569                                                 []   
4            dx_code_90172                             [procedure_code_30555]   

                                     MEDICATION_CODE  \
0  [medication_code_2350, medication_code_8630, m...   
1  [medication_code_7789, medication_code_3560, m...   
2  [medication_code_7982, medication_code_2452, m...   
3                             [medication_code_4036]   
4       [medication_code_6755, medication_code_5045]   

                                            LAB_CODE  \
0  [lab_code_8835, lab_code_9859, lab_code_9032, ...   
1                                                 []   
2  [lab_code_4198, lab_code_6603, lab_code_332, l...   
3                                                 []   
4                                    [lab_code_9112]   

                                               LABEL  
0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  
1                                       [1, 1, 1, 1]  
2                              [0, 0, 0, 0, 0, 0, 0]  
3                                                [0]  
4                                       [0, 0, 0, 0]

In [51]:
print("Rows x Columns in ehr_encounter_df = ", ehr_encounter_df.shape)
print("Total number of unique patients = ", len(ehr_encounter_df['PATIENT_ID'].unique()))
print("Total number of unique patients = ", ehr_encounter_df['PATIENT_ID'].nunique())

Rows x Columns in ehr_encounter_df =  (7222, 7)
Total number of unique patients =  878
Total number of unique patients =  878


#### Split at Patient Level

**We have to take care of the following condition:**
1. We have to split the dataset in such a way that the `train` and `test` partitions should have an 80:20 ratio of the **unique patients**. In other words, there should not be any overlapping encounters of a patient between the two partitions. It means, if total no. of unique patients in the original dataset is 878, then the unique patients in `train` will always be 702, and `test` will have 176 patients' records. 


**Note**: Interestingly, **the exact count of rows in each `train` and `test` partitions will change every time you run the function below.** The reason is that the split is based on `key='PATIENT_ID'`, due to which each time you run the function, it will allocate a **new set** of unique patients to train/test partition, where each patient has a different number of encounters. 

In [52]:
np.random.permutation(10)

array([1, 8, 9, 5, 6, 0, 7, 4, 3, 2])

In [53]:
DF.iloc[0]

A      37
B      10
C      43
D      55
N    hari
Name: 0, dtype: object

In [54]:
DF.iloc[range(10)]

A   B   C   D      N
0  37  10  43  55   hari
1  62  98  79  81  Shyam
2  56  73  25  49  Shyam
3  98  28   8  33    Ram
4  21  79  40  57   gita
5  60  35  56   7  Shyam
6  38  95  80   9   gita
7  98  38  17  66   hari
8  87  90  20  22  Shyam
9  89  81  64  87   hari

In [55]:
DF.iloc[np.random.permutation(10)]

A   B   C   D      N
4  21  79  40  57   gita
7  98  38  17  66   hari
8  87  90  20  22  Shyam
6  38  95  80   9   gita
2  56  73  25  49  Shyam
0  37  10  43  55   hari
3  98  28   8  33    Ram
5  60  35  56   7  Shyam
1  62  98  79  81  Shyam
9  89  81  64  87   hari

In [56]:
DF.loc[10:15,"A":"B"]

A   B
10  16  63
11  19  82
12  60  51
13  55  27
14  33  51
15  19  61

In [57]:
PATIENT_ID_FIELD = 'PATIENT_ID'
TEST_PERCENT = 0.2

In [58]:
def split_dataset_patient_level(df, key, test_percentage=0.2):
    
    df = df.iloc[np.random.permutation(len(df))]
    
    unique_values = df[key].unique()
    total_values = len(unique_values)
    
    sample_size = round(total_values * (1 - test_percentage ))
    
    
    train = df[df[key].isin(unique_values[:sample_size])].reset_index(drop=True)
    
    test = df[df[key].isin(unique_values[sample_size:])].reset_index(drop=True)
    
    print("Total number of unique patients in train = ", len(train['PATIENT_ID'].unique()))
    print("Total number of unique patients in test = ", len(test['PATIENT_ID'].unique()))
    print("Training partition has a shape = ", train.shape) 
    print("Test partition has a shape = ", test.shape)
    
    return train, test

In [59]:
train_df, test_df = split_dataset_patient_level(ehr_encounter_df, PATIENT_ID_FIELD, TEST_PERCENT)

Total number of unique patients in train =  702
Total number of unique patients in test =  176
Training partition has a shape =  (6822, 7)
Test partition has a shape =  (400, 7)


In [60]:
set([1,2,3]).intersection(set([3,4,5]))

{3}

In [61]:
assert len(set(train_df[PATIENT_ID_FIELD].unique()).intersection(set(test_df[PATIENT_ID_FIELD].unique()))) == 0
print("Test passed for patient data in only one partition")

Test passed for patient data in only one partition


In [62]:
assert (train_df[PATIENT_ID_FIELD].nunique()  + test_df[PATIENT_ID_FIELD].nunique()) == ehr_encounter_df[PATIENT_ID_FIELD].nunique()
print("Test passed for number of unique patients being equal!")

Test passed for number of unique patients being equal!


In [63]:
assert len(train_df)  + len(test_df) == len(ehr_encounter_df)
print("Test passed for number of total rows equal!")

Test passed for number of total rows equal!


#### Optional
- Check label distribution and use scikitlearn - https://scikit-learn.org/stable/auto_examples/model_selection/plot_cv_indices.html#sphx-glr-auto-examples-model-selection-plot-cv-indices-py

## Exercise  L3 - 3: Build Bucketed Numeric Feature with TF

### Instructions
- Given the Swiss heart disease dataset that we worked with earlier, build a bucketed numeric feature from the age feature. 
- For this exercise, use the Tensorflow csv function for loading the dataset directly into a TF tensor -https://www.tensorflow.org/api_docs/python/tf/data/experimental/make_csv_dataset. This approach will be useful for when you have much larger datasets and also allows you to bypass loading the dataset in Pandas.
- More information on the Tensorflow bucketized feature can be found here https://www.tensorflow.org/api_docs/python/tf/feature_column/bucketized_column. Bucketed features take as input the  numeric feature that we covered in the lesson. For the numeric feature, you do not need to normalize it like we did in the lesson. 

In [64]:
import tensorflow as tf
swiss_dataset_path = "./data/lesson_exercise_swiss_dataset.csv"
BATCH_SIZE =128
PREDICTOR_FIELD = 'num_label'

In [65]:
import pandas as pd
pd.read_csv(swiss_dataset_path).head()

age  num_label
0   32          1
1   34          1
2   35          3
3   36          1
4   38          2

In [66]:
next(iter([[1,2,3,4],[2,4,6,7]]))

[1, 2, 3, 4]

### Solution

In [67]:
# ETL with TF dataset make csv function
swiss_tf_dataset = tf.data.experimental.make_csv_dataset( swiss_dataset_path, batch_size=BATCH_SIZE, 
                                                         num_epochs=1, label_name=PREDICTOR_FIELD, header=True)

In [68]:
swiss_tf_dataset

<PrefetchDataset shapes: (OrderedDict([(age, (None,))]), (None,)), types: (OrderedDict([(age, tf.int32)]), tf.int32)>

In [69]:
swiss_dataset_batch = next(iter(swiss_tf_dataset))[0]
swiss_dataset_batch

OrderedDict([('age',
              <tf.Tensor: shape=(123,), dtype=int32, numpy=
              array([34, 55, 61, 54, 43, 51, 61, 43, 51, 73, 63, 43, 48, 47, 66, 58, 35,
                     53, 53, 38, 61, 56, 72, 62, 65, 61, 62, 63, 62, 69, 68, 60, 36, 59,
                     50, 65, 68, 55, 56, 51, 70, 59, 38, 60, 64, 53, 46, 65, 64, 62, 65,
                     53, 47, 56, 50, 70, 53, 53, 51, 32, 61, 42, 60, 52, 53, 60, 63, 64,
                     38, 56, 62, 55, 47, 51, 60, 56, 54, 42, 52, 66, 62, 52, 56, 51, 54,
                     50, 38, 61, 38, 50, 57, 61, 57, 40, 38, 63, 57, 57, 58, 59, 41, 46,
                     47, 55, 56, 58, 57, 60, 59, 61, 51, 59, 45, 74, 67, 56, 53, 61, 57,
                     63, 62, 69, 52], dtype=int32)>)])

In [70]:
# create TF numeric feature
tf_numeric_age_feature = tf.feature_column.numeric_column(key='age', default_value=0, dtype=tf.float64)

In [71]:
tf_numeric_age_feature

NumericColumn(key='age', shape=(1,), default_value=(0,), dtype=tf.float64, normalizer_fn=None)

In [72]:
#boundaries for the different age buckets
b_list = [ 0, 18, 25, 40, 55, 65, 80, 100]

#create TF bucket feature from numeric feature
tf_bucket_age_feature = tf.feature_column.bucketized_column(source_column=tf_numeric_age_feature, boundaries= b_list)

In [73]:
tf_bucket_age_feature

BucketizedColumn(source_column=NumericColumn(key='age', shape=(1,), default_value=(0,), dtype=tf.float64, normalizer_fn=None), boundaries=(0, 18, 25, 40, 55, 65, 80, 100))

In [74]:
def demo(feature_column, example_batch):
    
    
    feature_layer = tf.keras.layers.DenseFeatures(feature_column)
    
    
    print(feature_layer(example_batch))
    print("\nExample of one transformed row:")
    print(feature_layer(example_batch).numpy()[0])

In [75]:
print("Example bucket field:\n{}\n".format(tf_bucket_age_feature))

demo(tf_bucket_age_feature, swiss_dataset_batch)

Example bucket field:
BucketizedColumn(source_column=NumericColumn(key='age', shape=(1,), default_value=(0,), dtype=tf.float64, normalizer_fn=None), boundaries=(0, 18, 25, 40, 55, 65, 80, 100))

tf.Tensor(
[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 1. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]], shape=(123, 9), dtype=float32)

Example of one transformed row:
[0. 0. 0. 1. 0. 0. 0. 0. 0.]


## Exercise  L3 - 4:  Build Embedding Categorical Feature with TF

### Instructions
- Build a 10 dimension embedding feature for the PRINCIPAL_DIAGNOSIS_CODE field
- Here is the link to the Tensorflow Embedding column documentation -https://www.tensorflow.org/api_docs/python/tf/feature_column/embedding_column
- Some functions provided below to assist

In [77]:
ehr_line_df = pd.read_csv("./data/SYNTHETIC_EHR_DATASET.csv")
ehr_line_df.head()

ENCOUNTER_ID                     PATIENT_ID  \
0  udacity_health_encounter_id_1933  udacity_health_patient_id_817   
1  udacity_health_encounter_id_5664  udacity_health_patient_id_594   
2  udacity_health_encounter_id_1946  udacity_health_patient_id_169   
3  udacity_health_encounter_id_1528  udacity_health_patient_id_870   
4  udacity_health_encounter_id_7185  udacity_health_patient_id_722   

  PRINCIPAL_DIAGNOSIS_CODE        PROCEDURE_CODE       MEDICATION_CODE  \
0            dx_code_56522  procedure_code_20005                   NaN   
1            dx_code_39264                   NaN  medication_code_7471   
2            dx_code_83619                   NaN                   NaN   
3            dx_code_97434  procedure_code_31111                   NaN   
4            dx_code_68924                   NaN  medication_code_2250   

        LAB_CODE  LABEL  
0            NaN      0  
1            NaN      0  
2  lab_code_5311      0  
3            NaN      0  
4            NaN      0

In [78]:
cat_example_df = ehr_line_df[['ENCOUNTER_ID', 'PRINCIPAL_DIAGNOSIS_CODE', 'LABEL']]
cat_example_df.head()

ENCOUNTER_ID PRINCIPAL_DIAGNOSIS_CODE  LABEL
0  udacity_health_encounter_id_1933            dx_code_56522      0
1  udacity_health_encounter_id_5664            dx_code_39264      0
2  udacity_health_encounter_id_1946            dx_code_83619      0
3  udacity_health_encounter_id_1528            dx_code_97434      0
4  udacity_health_encounter_id_7185            dx_code_68924      0

In [79]:
#x = cat_example_df.pop('LABEL')
#cat_example_df.head()

In [80]:
#x

In [81]:
dataset = tf.data.Dataset.from_tensor_slices([[1, 2, 3],[4,5,6],[7,8,9]])
for element in dataset:
     print(element)

tf.Tensor([1 2 3], shape=(3,), dtype=int32)
tf.Tensor([4 5 6], shape=(3,), dtype=int32)
tf.Tensor([7 8 9], shape=(3,), dtype=int32)


In [82]:
type(dataset)

tensorflow.python.data.ops.dataset_ops.TensorSliceDataset

In [83]:
next(iter(dataset))

<tf.Tensor: shape=(3,), dtype=int32, numpy=array([1, 2, 3], dtype=int32)>

In [84]:
import numpy as np
DF = pd.DataFrame({"A": np.random.randint(1,100,1000),\
             "B": np.random.randint(1,100,1000),\
             "C": np.random.randint(1,100,1000),\
             "D": np.random.randint(1,100,1000),\
              "N": [np.random.choice(["Ram","Shyam","sita","gita","hari"]) for x in range(1000)] })

In [85]:
dict(DF)['A']

0      13
1      99
2      32
3      27
4      10
       ..
995    27
996    28
997    58
998     5
999    46
Name: A, Length: 1000, dtype: int64

In [86]:
#adapted from https://www.tensorflow.org/tutorials/structured_data/feature_columns
def df_to_dataset(df, predictor,  batch_size=32):
    df = df.copy()
    labels = df.pop(predictor)
    ds = tf.data.Dataset.from_tensor_slices((dict(df), labels))
    ds = ds.shuffle(buffer_size=len(df))
    ds = ds.batch(batch_size)
    return ds

In [87]:
BATCH_SIZE = 64
PREDICTOR_FIELD = 'LABEL'
categorical_tf_ds = df_to_dataset(cat_example_df, PREDICTOR_FIELD, batch_size=BATCH_SIZE)

In [88]:
# build vocab for categorical features
def write_vocabulary_file(vocab_list, field_name, default_value, vocab_dir='./vocab/'):
    output_file_path = os.path.join(vocab_dir, str(field_name) + "_vocab.txt")
    # put default value in first row as TF requires
    vocab_list = np.insert(vocab_list, 0, default_value, axis=0) 
    df = pd.DataFrame(vocab_list).to_csv(output_file_path, index=None, header=None)
    return output_file_path

def build_vocab_files(df, categorical_column_list, default_value='00'):
    vocab_files_list = []
    for c in categorical_column_list:
        v_file = write_vocabulary_file(df[c].unique(), c, default_value)
        vocab_files_list.append(v_file)
    return vocab_files_list

### Solution

In [89]:
import os
# add logic to add if not exist
#os.mkdir("./vocab/")

In [90]:
categorical_field_list = ["PRINCIPAL_DIAGNOSIS_CODE"]
vocab_files_list = build_vocab_files(cat_example_df, categorical_field_list)
vocab_files_list[0]

'./vocab/PRINCIPAL_DIAGNOSIS_CODE_vocab.txt'

In [91]:
with open(vocab_files_list[0],"r") as f:
    ct =0
    for line in f:
        print(line)
        if ct>5:
            break
        ct += 1

00

dx_code_56522

dx_code_39264

dx_code_83619

dx_code_97434

dx_code_68924

dx_code_85016



In [92]:
principal_diagnosis_vocab = tf.feature_column.categorical_column_with_vocabulary_file(
            key="PRINCIPAL_DIAGNOSIS_CODE", vocabulary_file = vocab_files_list[0], num_oov_buckets=1)

INFO:tensorflow:vocabulary_size = 6753 in PRINCIPAL_DIAGNOSIS_CODE is inferred from the number of elements in the vocabulary_file ./vocab/PRINCIPAL_DIAGNOSIS_CODE_vocab.txt.


In [93]:
principal_diagnosis_vocab

VocabularyFileCategoricalColumn(key='PRINCIPAL_DIAGNOSIS_CODE', vocabulary_file='./vocab/PRINCIPAL_DIAGNOSIS_CODE_vocab.txt', vocabulary_size=6753, num_oov_buckets=1, dtype=tf.string, default_value=-1)

In [94]:
dims = 10
cat_embedded = tf.feature_column.embedding_column(principal_diagnosis_vocab, dimension=dims)

In [95]:
categorical_tf_ds_batch = next(iter(categorical_tf_ds))[0]

In [96]:
demo(cat_embedded, categorical_tf_ds_batch)

tf.Tensor(
[[-5.10748804e-01 -8.61859843e-02 -1.16672233e-01  3.60501677e-01
   3.66676092e-01 -2.36660410e-02 -2.12401614e-01 -4.44710761e-01
   4.52660888e-01  7.61737004e-02]
 [-1.06796876e-01  6.48186877e-02 -4.32806671e-01  1.32008314e-01
   3.57566386e-01  3.71634029e-02 -2.21137881e-01  4.89579380e-01
   3.31925064e-01  1.07990038e-02]
 [ 1.41188219e-01  1.50117725e-01  9.28259566e-02 -3.82741868e-01
  -4.11248475e-01 -2.31902614e-01  2.59358436e-01  4.57047462e-01
   4.07946289e-01  3.59045565e-02]
 [-3.18739444e-01  6.19397283e-01 -4.89384755e-02 -3.20831358e-01
   5.67507923e-01 -2.17114925e-01 -2.73295134e-01  4.84817363e-02
   1.94014251e-01 -1.84793547e-01]
 [ 5.08643866e-01 -1.73768148e-01 -4.99874204e-01 -3.72543603e-01
  -1.23018198e-01  1.34180173e-01  1.67317525e-01 -1.29259527e-01
   7.70992190e-02  1.39450595e-01]
 [ 3.57710212e-01  2.69554198e-01  5.78429848e-02 -1.61504578e-02
  -2.92336911e-01 -3.66289765e-01  1.01332299e-01  3.78594291e-03
   5.66899359e-01 -1.3